### PR TITLE
Feature/1 edit config

### LINKED ISSUE #1: Implement edit-config functionality
See also cloudogu/cesapp-lib/issues/25

### PR DESCRIPTION
**Pull request by ppxl**
This PR resolves #1 by creating a CLI structure for edit-config which creates a port forward for each connection to the registry. The PF will be torn down after each respective registry call.

### CODE REVIEW

File: cmd/plugin/cli/dogu-config/config.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	errMsgDoguConfigServiceCreate = "cannot create config service: %w"
+)
+
+type defaultServiceFactory struct {
+	cliConfig configTransporter
+}
+
+func (sf *defaultServiceFactory) create(doguName string) (doguConfigService, error) {
+	k8sArgs := sf.cliConfig.Get(util.CliTransportParamK8sArgs)
+	cfg := (k8sArgs).(*genericclioptions.ConfigFlags)
```
Comment (cesmarvin) [inline cmd/plugin/cli/dogu-config/config.go]: 🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked

File: pkg/logger/formatter.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+	// CallerFirst - print caller info first
+	CallerFirst bool
+
+	// CustomCallerFormatter - set custom formatter for caller info
+	CustomCallerFormatter func(*runtime.Frame) string
+}
+
+// Format an log entry
+func (f *Formatter) Format(entry *logrus.Entry) ([]byte, error) {
```
Comment (cesmarvin) [inline pkg/logger/formatter.go:60]: 🚫 **[checkstyle]** <funlen> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Function 'Format' is too long (80 > 60)

File: pkg/logger/logger.go [hunk 3]
```diff
@@ -2,34 +2,186 @@ package logger
...
+	errorLevel int = iota
+	warningLevel
+	infoLevel
+	debugLevel
 )
 
-type Logger struct {
+var (
+	loggerInstance *NamedLogger
+	formatter      = createFormatter()
```
Comment (cesmarvin) [inline pkg/logger/logger.go:17]: 🚫 **[checkstyle]** <unused> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>var `formatter` is unused

File: pkg/logger/logger.go [hunk 4]
```diff
@@ -2,34 +2,186 @@ package logger
...
+			return file[len(file)-1] + ":" + function[len(function)-1] + ":" + strconv.Itoa(frame.Line) + " "
+		},
+	}
+
+	return formatter
+}
+
+func getLogLevelFromEnv() logrus.Level {
+	logLevelRef := viper.GetViper().Get(util.CliTransportLogLevel)
+	logLevel := *(logLevelRef).(*LogLevel)
```
Comment (cesmarvin) [inline pkg/logger/logger.go]: 🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked

File: pkg/plugin/dogu-config/service.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+type doguConfigService struct {
+	delegator delegator
+}
+
+// Edit opens an interactive dialogue to edit the value of the given key.
+func (s doguConfigService) Edit(registryKey string, deleteOnEmpty bool) error {
+	return s.delegator.Delegate(func(dogu *core.Dogu, editor doguConfigurationEditor) error {
+		matchingFields := matchConfigurationFields(dogu, registryKey)
+		if matchingFields == nil || len(matchingFields) == 0 {
```
Comment (cesmarvin) [inline pkg/plugin/dogu-config/service.go]: 🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1009: should omit nil check; len() for []github.com/cloudogu/cesapp-lib/core.ConfigurationField is defined as zero

Comment (cesmarvin) [inline pkg/plugin/dogu-config/service.go]: 🚫 **[checkstyle]** <gosimple> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>S1009: should omit nil check; len() for []github.com/cloudogu/cesapp-lib/core.ConfigurationField is defined as zero

File: cmd/plugin/cli/dogu-config/delete.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		Use:     "delete <dogu> <config-key>",
+		Long:    `Delete a given configuration key.`,
+		Aliases: []string{"d"},
+		Args:    cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			doguName := ""
+			configKey := ""
+
+			switch len(args) {
+			case 2:
```
Comment (ppxl) [inline cmd/plugin/cli/dogu-config/delete.go]: replace refactored `switch` with `if`?

File: cmd/plugin/cli/dogu-config/get.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		Use:     "get <dogu> <config-key>",
+		Long:    `Fetch a value for a given configuration key.`,
+		Aliases: []string{"g"},
+		Args:    cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			doguName := ""
+			configKey := ""
+
+			switch len(args) {
+			case 2:
```
Comment (ppxl) [inline cmd/plugin/cli/dogu-config/get.go]: replace `switch` with  `if`?

File: cmd/plugin/cli/dogu-config/list.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+func listCmd(factory serviceFactory) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "list <dogu>",
+		Long:    `Fetch a list of possible configuration keys.`,
+		Aliases: []string{"l", "ls"},
+		Args:    cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			doguName := ""
+			switch len(args) {
+			case 1:
```
Comment (ppxl) [inline cmd/plugin/cli/dogu-config/list.go]: replace `switch` with `if`?

File: cmd/plugin/cli/dogu-config/set.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+		Long:    `Create or update a value for a given configuration key.`,
+		Aliases: []string{"s"},
+		Args:    cobra.ExactArgs(3),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			doguName := ""
+			configKey := ""
+			configValue := ""
+
+			switch len(args) {
+			case 3:
```
Comment (ppxl) [inline cmd/plugin/cli/dogu-config/set.go]: replace `switch` with `if`?

File: docs/development/development_en.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+## Encouraged Command Structure
+
+The used CLI framework [github.com/spf13/cobra](https://github.com/spf13/cobra) provides a handy convention on how to structure a CLI commands:
+> APPNAME VERB NOUN --ADJECTIVE
+
+Anyhow, this plug-in will most likely deal with a large enough bandwidth of use cases for this plugin to enhance this command structure. To provide a matching narrative for different use cases the structure adds a GROUP command:
+> kubectl ces GROUP VERB NOUN --ADJECTIVE
+
+This enables to keep an easily understandable command structure and enables use cases like these (some may not be implemented yet):
```
Comment (ppxl) [inline docs/development/development_en.md]:
```suggestion
This still keeps the command structure easily understandable and enables use cases like these (some may not be implemented yet):
```

File: go.mod [hunk 11]
```diff
@@ -1,19 +1,25 @@
 module github.com/cloudogu/kubectl-ces-plugin
 
-go 1.18
+go 1.20
 
 require (
-	github.com/fatih/color v1.14.1
+	github.com/cloudogu/cesapp-lib v0.0.0-20230222095820-819178066686
```
Comment (ppxl) [inline go.mod]: FIXME

File: pkg/keys/interfaces.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	//	DoguRegistry() registry.DoguRegistry
+	//	// BlueprintRegistry to maintain a blueprint history
+	//	BlueprintRegistry() configurationContext
+	//	// RootConfig returns a WatchConfigurationContext for the root context
+	//	RootConfig() registry.WatchConfigurationContext
+	//	// GetNode returns all keys that are included in any path, packed as Node
+	//	GetNode() (registry.Node, error)
+	// }
+	// ```
+	// Sadly, Go is too stoopid to do recursive duck-typing.
```
Comment (ppxl) [inline pkg/keys/interfaces.go]:
```suggestion
```

File: pkg/logger/logger.go [hunk 13]
```diff
@@ -2,34 +2,128 @@ package logger
...
+	namedLog := &NamedLogger{logger: log, name: "throwAway"}
+	loggerInstance = namedLog
+	core.GetLogger = func() core.Logger {
+		return namedLog
+	}
 }
 
-func NewLogger() *Logger {
-	return &Logger{}
+func GetInstance() *NamedLogger {
```
Comment (ppxl) [inline pkg/logger/logger.go:35]: add doc comment

File: pkg/logger/level.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+
+	"github.com/cloudogu/kubectl-ces-plugin/cmd/plugin/cli/util"
+)
+
+// LogLevelKey identifies
```
Comment (ppxl) [inline pkg/logger/level.go]: fix doc comment

File: go.mod (hunk 11 above)
Comment (ppxl) [reply go.mod]: fixed

**Pull request merged**